sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -257,8 +257,16 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
     }
     // The inferred schema may have different field names as the table schema, we should respect
     // it, but also respect the exprId in table relation output.
-    assert(result.output.length == relation.output.length &&
-      result.output.zip(relation.output).forall { case (a1, a2) => a1.dataType == a2.dataType })
+    assert(result.output.length == relation.output.length,
+      s"Target table has ${result.output.length} columns, " +
+        s"but source table has ${relation.output.length} columns. " +
+        s"It may need to recreate the table ${relation.tableMeta.identifier}")
+    result.output.zip(relation.output).foreach { case (a1, a2) =>
+      assert(a1.dataType.sameType(a2.dataType),
+        s"Data type of column ${a1.name} in target table is ${a1.dataType.typeName}, " +
+          s"but column ${a2.name} in source table is ${a2.dataType.typeName}. " +
+          s"It may need to recreate the table ${relation.tableMeta.identifier}")
+    }
```

Comment:
Thank you @LantaoJin Could we add a test case?

Reply:
`spark.sql.sources.*` can not add in SQL of `CREATE TABLE` in current Spark version. Any other way to go to this invalid code path?

Reply:
`sparkSession.metadataHive.runSqlHive`?
https://github.com/apache/spark/blob/96179732aad28418d486de1a365cd5d68c3db910/sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveMetastoreCatalogSuite.scala#L314-L320